### PR TITLE
Nested tensor forward only chunk operations

### PR DESCRIPTION
**Pull request by drisspg**
# Summary

Taking over this pr: https://github.com/pytorch/pytorch/pull/83736

Adding support for chunk without autograd support

### CODE REVIEW

**Review by cpuhrsch — COMMENTED** [review]
So what happens why I try to use autograd together with these ops?

File: aten/src/ATen/native/TensorShape.cpp
Comment (cpuhrsch) [inline aten/src/ATen/native/TensorShape.cpp:2178]: Why was this removed?

Comment (drisspg) [reply aten/src/ATen/native/TensorShape.cpp:2178]: whoops I think this comment was added after Joels pr I will add back

**Comment by drisspg** [comment]
> So what happens why I try to use autograd together with these ops?

For ops split, split_with_sizes since the registered derivative formula calls `self.size()` to save for backward we get an error:

```
import torch
a = torch.nested_tensor([torch.randn(2, 3, 6, requires_grad=True), torch.randn(4, 5, 6, requires_grad=True)])
b = torch.split(a, 2, dim=-1)

RuntimeError: Internal error: NestedTensorImpl doesn't support sizes. Please file an issue on https://github.com/pytorch/nestedtensor
```

For the composite ops like, chunk, narrow even though they are composite the same error arises because they all end up hitting split or split with sizes that has the `.sizes` call.

**Comment by drisspg** [comment]
> > So what happens why I try to use autograd together with these ops?
> 
> For ops split, split_with_sizes since the registered derivative formula calls `self.size()` to save for backward we get an error:
> 
> ```
> import torch
> a = torch.nested_tensor([torch.randn(2, 3, 6, requires_grad=True), torch.randn(4, 5, 6, requires_grad=True)])
> b = torch.split(a, 2, dim=-1)
> 
> RuntimeError: Internal error: NestedTensorImpl doesn't support sizes. Please file an issue on https://github.com/pytorch/nestedtensor
> ```
> 
> For the composite ops like, chunk, narrow even though they are composite the same error arises because they all end up hitting split or split with sizes that has the `.sizes` call.

@cpuhrsch 
Follow up pr would be to use the Mikayla's derivative.yaml notation to register the derivative with slice and split/split_with_sizes.

**Comment by cpuhrsch** [comment]
@drisspg - let's register "NotImplemented" as the derivative in a follow up. Let's at least have a good error message.

File: test/test_nestedtensor.py
Comment (cpuhrsch) [inline test/test_nestedtensor.py:606]: Just for future reference, this doesn't *have* to fail. We could easily narrow along an irregular dimension as long as there's enough data, it'll just become regular as a result.

**Review by cpuhrsch — APPROVED** [review]
Neat!

File: test/test_nestedtensor.py
Comment (mikaylagawarecki) [inline test/test_nestedtensor.py:633]: @drisspg This might landrance with the nested_tensor/as_nested_tensor PR )":

Would you be open to waiting for that to land before landing this? Asking because I'm landing that via phabricator due to internal changes that will break tests and phabricator tests take extremely long to run (and are almost done for the latest rebase right now!)

Alternatively, could we land this test case specifically in a separate PR (I could open a PR that adds this test back after my PR lands)? If not I'll wait for this to land and then rebase on my end (but just a little nervous about getting everything in hahah)

Comment (drisspg) [reply test/test_nestedtensor.py:633]: Yeah of course, I just canceled my merge job. Just Ping me and let mw know when I can land. I assume it will be tomorrow after the diff train with yours lands?

Comment (mikaylagawarecki) [reply test/test_nestedtensor.py:633]: hm mine is not in diff train bcos i'm clicking land on phabricator! If nothing goes wrong it should be landed before tmr morning, will ping you!

File: aten/src/ATen/native/nested/NestedTensorMath.cpp
Comment (drisspg) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:661]: I included the updates to this clone function here

File: aten/src/ATen/native/native_functions.yaml
Comment (drisspg) [inline aten/src/ATen/native/native_functions.yaml:1283]: @cpuhrsch  when building from source with debug=0:

```
TORCH_INTERNAL_ASSERT(
          aliased_input.storage().is_alias_of(aliased_output.storage()),
            op_name);
````
The above line will be added to the codgened kernel for nested tensor because of this signature. I added a comment on the test. I really don't like this because it makes dev experience worse but I couldn't figure out a nice way around it.

**Comment by drisspg** [comment]
@cpuhrsch removed all by the restricted chunk and the small update to clone

**Review by cpuhrsch — APPROVED** [review]

**Pull request closed without merging**
